File: frontend/src/api/traffic.ts
import { ApiError, type RouteSummary } from './routing'

export type TrafficProfile = {
  id: string
  name: string
  version: string
  source_name: string
  source_window: string
  period: 'weekday_morning' | 'weekday_afternoon'
  graph_version: string
  observation_count: number
  volume_observation_count: number
  matched_edge_count: number
  median_multiplier: number
  p85_multiplier: number
  p90_multiplier: number
  p95_multiplier: number
  created_at: string
}

export type TrafficDataQuality = {
  row_count: number
  accepted_count: number
  rejected_count: number
  matched_station_count: number
  unmatched_station_count: number
  missing_speed_percent: number
  missing_volume_percent: number
  quality_flags: string[]
}

export type TrafficImportResponse = {
  import_id: string
  source_name: string
  raw_sha256: string
  normalized_path: string
  quality: TrafficDataQuality
  profiles: TrafficProfile[]
}

export type PortalHighway = {
  id: number
  name: string
  direction: string
  station_count: number
}

export type PortalAcquireSettings = {
  startDate: string
  endDate: string
  highwayIds: number[]
  resolution: '00:15:00' | '01:00:00'
}

export type PortalAcquireResponse = {
  requested_highways: PortalHighway[]
  downloaded_observation_count: number
  normalized_station_count: number
  import_result: TrafficImportResponse
}

export type ReliabilitySettings = {
  planningMode: 'arrive_by' | 'depart_at'
  arrivalDeadline: string
  bufferMinutes: number
  confidenceTarget: number
  sampleCount: number
  profileId: string | null
}

export type ReliabilityResponse = {
  planning_mode: 'arrive_by' | 'depart_at'
  evidence_level: 'modeled_uncalibrated' | 'historically_calibrated'
  mean_seconds: number
  median_seconds: number
  p85_seconds: number
  p90_seconds: number
  p95_seconds: number
  likely_low_seconds: number
  likely_high_seconds: number
  on_time_probability: number | null
  latest_safe_departure: string | null
  planned_departure_time: string
  median_arrival_time: string
  p85_arrival_time: string
  p90_arrival_time: string
  p95_arrival_time: string
  confidence_arrival_time: string
  confidence_target: number
  sample_count: number
  source_name: string
  source_window: string
  profile_version: string
  early_departure_benefits: {
    minutes_earlier: 5 | 10 | 15
    on_time_probability: number
    improvement: number
  }[]
  assumptions: string[]
}

export async function listTrafficProfiles(signal?: AbortSignal): Promise<TrafficProfile[]> {
  const response = await fetch('/api/traffic/profiles', { signal })
  const payload = await parseResponse<{ profiles: TrafficProfile[] }>(response)
  return payload.profiles
}

export async function importTraffic(
  file: File,
  sourceName: string,
): Promise<TrafficImportResponse> {
  const form = new FormData()
  form.append('file', file)
  form.append('source_name', sourceName)
  return parseResponse(
    await fetch('/api/traffic/import', { method: 'POST', body: form }),
  )
}

export async function listPortalHighways(signal?: AbortSignal): Promise<PortalHighway[]> {
  const response = await fetch('/api/traffic/portal/highways', { signal })
  const payload = await parseResponse<{ highways: PortalHighway[] }>(response)
  return payload.highways
}

export async function acquirePortalTraffic(
  settings: PortalAcquireSettings,
): Promise<PortalAcquireResponse> {
  return parseResponse(
    await fetch('/api/traffic/portal/acquire', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        start_date: settings.startDate,
        end_date: settings.endDate,
        highway_ids: settings.highwayIds,
        resolution: settings.resolution,
        days_of_week: [2, 3, 4, 5, 6],
      }),
    }),
  )
}

export async function simulateReliability(
  route: RouteSummary,
  departureTime: string,
  settings: ReliabilitySettings,
): Promise<ReliabilityResponse> {
  return parseResponse(
    await fetch('/api/simulations', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        route,
        planning_mode: settings.planningMode,
        departure_time: new Date(departureTime).toISOString(),
        arrival_deadline:
          settings.planningMode === 'arrive_by'
            ? new Date(settings.arrivalDeadline).toISOString()
            : null,
        buffer_minutes: settings.bufferMinutes,
        confidence_target: settings.confidenceTarget,
        sample_count: settings.sampleCount,
        profile_id: settings.profileId,
      }),
    }),
  )
}

async function parseResponse<T>(response: Response): Promise<T> {
  if (response.ok) return response.json() as Promise<T>
  const payload = (await response.json().catch(() => null)) as {
    detail?: { code?: string; message?: string }
  } | null
  throw new ApiError(
    payload?.detail?.code ?? 'request_failed',
    payload?.detail?.message ?? 'The traffic request failed.',
  )
}

File: backend/app/schemas/traffic.py
"""Traffic import, profile, and reliability API schemas."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from backend.app.schemas.routing import RouteSummary


class TrafficDataQuality(BaseModel):
    row_count: int = Field(ge=0)
    accepted_count: int = Field(ge=0)
    rejected_count: int = Field(ge=0)
    matched_station_count: int = Field(ge=0)
    unmatched_station_count: int = Field(ge=0)
    missing_speed_percent: float = Field(ge=0, le=100)
    missing_volume_percent: float = Field(ge=0, le=100)
    quality_flags: list[str]


class TrafficProfile(BaseModel):
    id: UUID
    name: str
    version: str
    source_name: str
    source_window: str
    period: Literal["weekday_morning", "weekday_afternoon"]
    graph_version: str
    observation_count: int = Field(ge=1)
    volume_observation_count: int = Field(default=0, ge=0)
    matched_edge_count: int = Field(default=0, ge=0)
    median_multiplier: float = Field(ge=1)
    p85_multiplier: float = Field(ge=1)
    p90_multiplier: float = Field(ge=1)
    p95_multiplier: float = Field(ge=1)
    created_at: datetime


class TrafficProfilesResponse(BaseModel):
    profiles: list[TrafficProfile]


class TrafficImportResponse(BaseModel):
    import_id: UUID
    source_name: str
    raw_sha256: str
    normalized_path: str
    quality: TrafficDataQuality
    profiles: list[TrafficProfile]


class ReliabilityRequest(BaseModel):
    route: RouteSummary
    planning_mode: Literal["arrive_by", "depart_at"] = "arrive_by"
    departure_time: datetime
    arrival_deadline: datetime | None = None
    buffer_minutes: int = Field(default=8, ge=0, le=180)
    confidence_target: float = Field(default=0.9, ge=0.5, le=0.99)
    sample_count: int = Field(default=1000, ge=100, le=10000)
    profile_id: UUID | None = None

    @model_validator(mode="after")
    def validate_times(self) -> "ReliabilityRequest":
        if not _aware(self.departure_time):
            raise ValueError("Departure time must include a timezone.")
        if self.planning_mode == "arrive_by":
            if self.arrival_deadline is None or not _aware(self.arrival_deadline):
                raise ValueError("Arrive-by planning requires a timezone-aware deadline.")
            if self.arrival_deadline <= self.departure_time:
                raise ValueError("Arrival deadline must be after departure time.")
        return self


class EarlyDepartureBenefit(BaseModel):
    minutes_earlier: Literal[5, 10, 15]
    on_time_probability: float = Field(ge=0, le=1)
    improvement: float = Field(ge=0, le=1)


class ReliabilityResponse(BaseModel):
    planning_mode: Literal["arrive_by", "depart_at"]
    evidence_level: Literal["modeled_uncalibrated", "historically_calibrated"]
    mean_seconds: float = Field(ge=0)
    median_seconds: float = Field(ge=0)
    p85_seconds: float = Field(ge=0)
    p90_seconds: float = Field(ge=0)
    p95_seconds: float = Field(ge=0)
    likely_low_seconds: float = Field(ge=0)
    likely_high_seconds: float = Field(ge=0)
    on_time_probability: float | None = Field(default=None, ge=0, le=1)
    latest_safe_departure: datetime | None = None
    planned_departure_time: datetime
    median_arrival_time: datetime
    p85_arrival_time: datetime
    p90_arrival_time: datetime
    p95_arrival_time: datetime
    confidence_arrival_time: datetime
    confidence_target: float
    sample_count: int
    source_name: str
    source_window: str
    profile_version: str
    early_departure_benefits: list[EarlyDepartureBenefit]
    assumptions: list[str]


def _aware(value: datetime) -> bool:
    return value.tzinfo is not None and value.utcoffset() is not None
